scalding-core/src/main/scala/com/twitter/scalding/ExecutionOptimizationRules.scala

```diff
@@ -0,0 +1,89 @@
+    // to prove this is safe, see that if you have
+    // LiteralExecution[_ <: T] we can call .evaluate to get
+    // Execution[_ <: T] which due to covariance is
+    // Execution[T], and then using toLiteral we can get
+    // LiteralExecution[T]
+    //
+    // that would be wasteful to apply since the final
+    // result is identity.
+    l.asInstanceOf[LiteralExecution[T]]
+  }
+
+  def toLiteral: FunctionK[Execution, LiteralExecution] =
+    Memoize.functionK[Execution, LiteralExecution](
+      new Memoize.RecursiveK[Execution, LiteralExecution] {
+        override def toFunction[A] = {
+          case (e @ Execution.ReaderExecution, _) =>
+            Literal.Const(e)
+          case (e: Execution.FutureConst[a], _) =>
+            Literal.Const(e)
+          case (e: Execution.UniqueIdExecution[a], _) =>
+            Literal.Const(e)
+          case (e: Execution.FlowDefExecution, _) =>
+            Literal.Const(e)
+          case (e: Execution.WriteExecution[a], _) =>
+            Literal.Const(e)
+          case (e: Execution.GetCounters[a], f) =>
+            widen(Literal.Unary[Execution, a, (a, ExecutionCounters)](f(e.prev), Execution.GetCounters(_: Execution[a])))
+          case (e: Execution.ResetCounters[a], f) =>
+            Literal.Unary(f(e.prev), Execution.ResetCounters(_: Execution[a]))
+          case (e: Execution.WithNewCache[a], f) =>
+            Literal.Unary(f(e.prev), Execution.WithNewCache(_: Execution[a]))
+          case (e: Execution.TransformedConfig[a], f) =>
+            Literal.Unary(f(e.prev), Execution.TransformedConfig(_: Execution[a], e.fn))
+          case (e: Execution.OnComplete[a], f) =>
+            Literal.Unary(f(e.prev), Execution.OnComplete(_: Execution[a], e.fn))
+          case (e: Execution.RecoverWith[a], f) =>
+            Literal.Unary(f(e.prev), Execution.RecoverWith(_: Execution[a], e.fn))
+          case (e: Execution.Mapped[a, b], f) =>
+            Literal.Unary(f(e.prev), Execution.Mapped(_: Execution[a], e.fn))
+          case (e: Execution.FlatMapped[a, b], f) =>
+            Literal.Unary(f(e.prev), Execution.FlatMapped(_: Execution[a], e.fn))
+          case (e: Execution.Zipped[a, b], f) =>
+            Literal.Binary(f(e.one), f(e.two), Execution.Zipped(_: Execution[a], _: Execution[b]))
+        }
+      }
+    )
+
+  object ZipWrite extends PartialRule[Execution] {
+    override def applyWhere[T](on: Dag[Execution]) = {
+      /*
+      * run this and that in parallel, without any dependency. This will
+      * be done in a single cascading flow if possible.
+      *
+      * If both sides are write executions then merge them
+      */
+      case Execution.Zipped(Execution.WriteExecution(oH, oT, oR), Execution.WriteExecution(tH, tT, tR)) =>
+        val newFn = { tup: (Config, Mode, Execution.Writer, ConcurrentExecutionContext) =>
+          Execution.failFastZip(oR(tup), tR(tup))(tup._4)
+        }
+        Execution.WriteExecution(oH, tH :: tT ::: oT, newFn)
+    }
```

Comment:
I think we want a few more rules:

1. pushing `Execution.Mapped(Execution.WriteExecution` into the `WriteExecution` function.
2. `zip(a.map(f), b)` == `zip(a, b).map { case (left, right) => (f(left), right) }` similarly on the right hand side.
3. there are some internal flatMaps that we can have case class functions to pattern match on optimize in some ways, e.g the random UUID ones, we are better doing `a.flatMap(b).zip(c)` to `a.flatMap { x => b(x).zip(c) }` if we know that `a` is cheap to compute. Probably all our FromFuture examples are like this by assumption, so we don't want to kick off any hadoop jobs until the futures we depend on succeed.

Does this make sense?

Reply:
done.